core/jobrunner.py

```diff
@@ -92,11 +92,95 @@ class JobRunner(metaclass=_MetaRunner):
     A |JobRunner| instance can be passed to |run| with a keyword argument ``jobrunner``. If this argument is omitted, the instance stored in ``config.default_jobrunner`` is used.
     """
 
-    def __init__(self, parallel=False, maxjobs=0, maxthreads=256):
+    def __init__(self, parallel: bool = False, maxjobs: int = 0, maxthreads: int = 256):
+        # Initialize protected attributes
+        self._parallel = False
+        self._maxjobs = 0
+        self._job_limit = None
+        self._maxthreads = 0
+        self._jobthread_limit = None
+
+        # Set properties
         self.parallel = parallel
         self.maxjobs = maxjobs
-        self.semaphore = threading.BoundedSemaphore(maxjobs) if maxjobs else None
-        self._jobthread_limit = threading.BoundedSemaphore(maxthreads) if maxthreads else None
+        self.maxthreads = maxthreads
+
+    @property
+    def parallel(self) -> bool:
+        """
+        Whether jobs can be run simultaneously or not. Defaults to ``False``.
+
+        When set to ``True``, jobs can be run in parallel, with the level of parallelization determined by the properties :attr:`~scm.plams.core.jobrunner.JobRunner.maxjobs` and :attr:`~scm.plams.core.jobrunner.JobRunner.maxthreads`.
+        """
+        return self._parallel
+
+    @parallel.setter
+    def parallel(self, value: bool) -> None:
+        self._parallel = value
+
+    @property
+    def maxjobs(self) -> int:
+        """
+        Maximum number of jobs which can be run in parallel. Defaults to ``0``.
+
+        When set to ``0``, no job limit is enforced.
+
+        Note that to take effect, this requires :attr:`~scm.plams.core.jobrunner.JobRunner.parallel` to be ``True``.
+        The number of jobs may also be limited by :attr:`~scm.plams.core.jobrunner.JobRunner.maxthreads`.
+
+        When jobs are running using this, updating this value will block until its usage is complete.
+        """
+        return self._maxjobs
+
+    @maxjobs.setter
```

Comment:
N.B. as per PR description, could also just make these readonly and not have the option of changing outside of initialisation